connect/src/com/telenor/mobileconnect/ui/OperatorSelectionActivity.java

```diff
@@ -0,0 +1,108 @@
+
+    static final String MCC_MNC_PARAM_NAME = "mcc_mnc";
+    static final String MCC_MNC_PARAM_SEP = "_";
+    static final String SUBSCRIBER_ID_QUERY_PARAM_NAME = "subscriber_id";
+
+    @SuppressLint("MissingSuperCall")
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState, OperatorSelectionWebFragment.class);
+    }
+
+    @Override
+    public void onSuccess(Object successData) {
+
+        Uri uri = (Uri) successData;
+        String[] mccAndMnc = uri.getQueryParameter(MCC_MNC_PARAM_NAME).split(MCC_MNC_PARAM_SEP);
+
+        @SuppressWarnings("unchecked")
+        Map<String, String> parameters = (Map<String, String>) this
+                .getIntent()
+                .getExtras()
+                .get(ConnectUtils.PARAM_SNAPHOT_EXTRA);
+        @SuppressWarnings("unchecked")
+        List<String> uiLocales = (List<String>) this
+                .getIntent()
+                .getExtras()
+                .get(ConnectUtils.UI_LOCALES_SNAPHOT_EXTRA);
+        final int requestCode = this
+                .getIntent()
+                .getExtras()
+                .getInt(ConnectUtils.REQUEST_CODE_EXTRA);
+        final int customLoadingLayout = this
+                .getIntent()
+                .getExtras()
+                .getInt(ConnectUtils.CUSTOM_LOADING_SCREEN_EXTRA, ConnectSdk.NO_CUSTOM_LAYOUT);
+        MobileConnectSdkProfile sdkProfile = (MobileConnectSdkProfile) ConnectSdk.getSdkProfile();
+        sdkProfile.onStartAuthorization(
+                parameters,
+                uiLocales,
+                mccAndMnc[0],
+                mccAndMnc[1],
+                uri.getQueryParameter(SUBSCRIBER_ID_QUERY_PARAM_NAME),
+                new SdkProfile.OnStartAuthorizationCallback() {
+                    @Override
+                    public void onSuccess(Intent nextIntent) {
+                        if (!isConnectActivity(nextIntent)) {
+                            throw new RuntimeException("Wrong intent.");
+                        }
+                        Activity activity = OperatorSelectionActivity.this;
+                        if (customLoadingLayout != ConnectSdk.NO_CUSTOM_LAYOUT) {
+                            nextIntent.putExtra(
+                                    ConnectUtils.CUSTOM_LOADING_SCREEN_EXTRA,
+                                    customLoadingLayout);
+                        }
+                        activity.startActivityForResult(nextIntent, requestCode);
+                    }
+
+                    @Override
+                    public void onError() {
+                        cancelAndShowErrorMessage(OperatorSelectionActivity.this);
+                    }
+                });
+    }
+
+
+
+    @Override
+    public void onError(Object errorData) {
+        cancelAndShowErrorMessage(this);
+    }
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
```

Comment:
No, I just don't think that the OperatorSelection activity has to receive tokens. In Android typically the logic typically is, if you need to fetch some info (let's say contact) you will launch ContactActivity, (the user selects it) you set the result there, call finish() and handle the result in the original the activity that asked for the contact.

In your case, the way u've implemented it looks like OperatorSelection asks ConnectActivity for tokens, and receives them. Which is wrong, right? The way I see things is since ConnectActivity needs an authorization_url to proceed, it either have it (in case of Mobile Connect) or it need to fetches it. In the second case, it'll call OperatorSelection activity for help.

In pseudo code, something like:

ConnectSdk::authenticate always calls app's signinactivity.startforresult(connect_activity)
   
 connect_activity: onStart -> 1) if authorization_url is known (that is if mcc/mnc are present or if it is ConnectId) -> proceed as usual;
                               else 2) connect_activity.startforresult(os_selection)

in case of 2) , os_selection does it job, call finish(). and then ->

    connect_activity: result ->called from os_selection' finish, grab the authorization_url and proceed as in the previous case